#Pix2Text

In [ ]:
import torch
torch.cuda.empty_cache()

In [ ]:
!pip install -qU langchain jq tiktoken langchain_community langchain_chroma langchain-huggingface huggingface-hub sentence_transformers
!pip install einops datasets bitsandbytes accelerate peft flash_attn

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 7.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 54.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 737.4/737.4 kB 46.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 59.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 84.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 417.5/417.5 kB 35.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 249.1/249.1 kB 25.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 559.5/559.5 kB 43.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 88.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.0/94.0 kB 10.4 MB/s eta 0:

In [ ]:
!pip uninstall -y transformers


Found existing installation: transformers 4.44.2
Uninstalling transformers-4.44.2:
  Successfully uninstalled transformers-4.44.2


In [ ]:
!pip install transformers
!pip install sklearn
!pip install wandb
!pip install logging
!pip install --upgrade torch
!nvidia-smi

  Using cached transformers-4.44.2-py3-none-any.whl.metadata (43 kB)
Using cached transformers-4.44.2-py3-none-any.whl (9.5 MB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 76.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.3/207.3 kB 20.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.4/311.4 kB 29.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.0/96.0 kB 8.5 MB/s et

Sun Sep 15 07:56:25 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA L4                      Off | 00000000:00:03.0 Off |                    0 |
| N/A   41C    P8              16W /  72W |      1MiB / 23034MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

# Langchain

In [ ]:
from langchain_huggingface import ChatHuggingFace, HuggingFacePipeline, HuggingFaceEmbeddings
from langchain_community.embeddings import HuggingFaceInstructEmbeddings, HuggingFaceBgeEmbeddings, HuggingFaceHubEmbeddings, HuggingFaceHubEmbeddings
from langchain_huggingface import HuggingFacePipeline

/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:132: UserWarning: Field "model_name" in HuggingFaceInferenceAPIEmbeddings has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(


In [ ]:
import bs4, getpass, os, tiktoken
from langchain import hub
from langchain_chroma import Chroma
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import TextLoader
from langchain.prompts import ChatPromptTemplate
from transformers import BitsAndBytesConfig

# Load, chunk and index the contents of the blog.
#loader = TextLoader("/content/output/output.md")
#docs = loader.load()

def num_tokens_from_string(string: str, encoding_name: str) -> int:
    """Returns the number of tokens in a text string."""
    encoding = tiktoken.get_encoding(encoding_name)
    num_tokens = len(encoding.encode(string))
    return num_tokens
#num_tokens_from_string(question, "cl100k_base")
def cosine_similarity(vec1, vec2):
    dot_product = np.dot(vec1, vec2)
    norm_vec1 = np.linalg.norm(vec1)
    norm_vec2 = np.linalg.norm(vec2)
    return dot_product / (norm_vec1 * norm_vec2)
#similarity = cosine_similarity(query_result, document_result)

In [ ]:
model_name = "BAAI/bge-small-en"
model_kwargs = {"device": "cuda:0"}
encode_kwargs = {"normalize_embeddings": True}
hf = HuggingFaceBgeEmbeddings(
    model_name=model_name, model_kwargs=model_kwargs, encode_kwargs=encode_kwargs
)
#document = pages

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/90.8k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/684 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/133M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/366 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/711k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
#TEST IGNORE
num_tokens_from_string(document, "cl100k_base")
query_result = hf.embed_query(question)
document_result = hf.embed_query(docs)
len(query_result)

AttributeError: 'list' object has no attribute 'replace'

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
vectorstore = Chroma(persist_directory="/content/drive/MyDrive/UWA/Sem 4/Capstone/Project/vector",embedding_function=hf)

In [ ]:
retriever = vectorstore.as_retriever(search_kwargs={"k": 3})

In [ ]:

def format_docs(docs):
    return "\n\n".join(" page number : "+ str(doc.metadata['page']) +" "+ doc.page_content for doc in docs)

changing the format docs function to include the page number as part of the content for each chunk, not only in the metadata.

In [ ]:
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

# Load from Hugging Face hub
model = AutoModelForSeq2SeqLM.from_pretrained("ShilpaSandhya/phi3_lora_chemical_eng")
tokenizer = AutoTokenizer.from_pretrained("ShilpaSandhya/phi3_lora_chemical_eng")

`low_cpu_mem_usage` was None, now set to True since model is quantized.


Loading checkpoint shards:   0%|          | 0/6 [00:00<?, ?it/s]

In [ ]:
model_name = "microsoft/Phi-3-medium-4k-instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)

quantization_config = BitsAndBytesConfig(load_in_4bit=True,
                                bnb_4bit_quant_type='nf4',
                                bnb_4bit_compute_dtype='float16',
                                #bnb_4bit_compute_dtype=torch.bfloat16,
                                bnb_4bit_use_double_quant=True)
model = AutoModelForCausalLM.from_pretrained(model_name, quantization_config=quantization_config, attn_implementation="flash_attention_2", trust_remote_code=True)

In [ ]:
!pip install transformers accelerate

In [ ]:
# https://dev.to/debapriyadas/cloning-and-running-llama-31-model-from-hugging-face-using-python-3m80
import transformers
import torch

model_id = "meta-llama/Meta-Llama-3.1-8B"

pipeline = transformers.pipeline(
    "text-generation",
    model=model_id,
    model_kwargs={"torch_dtype": torch.bfloat16},
    device_map="auto",
)

messages = [
    {"role": "system", "content": "You are a pirate chatbot who always responds in pirate speak!"},
    {"role": "user", "content": "Who are you?"},
]

outputs = pipeline(
    messages,
    max_new_tokens=256,
)
print(outputs[0]["generated_text"][-1])

OSError: You are trying to access a gated repo.
Make sure to have access to it at https://huggingface.co/meta-llama/Meta-Llama-3.1-8B.
401 Client Error. (Request ID: Root=1-66e6a428-3aea76e7398725100c11968a;b33ff753-55c9-4850-905a-2ff1c0ac3708)

Cannot access gated repo for url https://huggingface.co/meta-llama/Meta-Llama-3.1-8B/resolve/main/config.json.
Access to model meta-llama/Meta-Llama-3.1-8B is restricted. You must have access to it and be authenticated to access it. Please log in.

Earlier we were using phi 3 mini which couldnt use complex prompts involving page number. after switching to phi 3 medium with flash attention 2, it can understand the prompt template better and extract page nmber and chapter number from chunks.

In [ ]:
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer, max_new_tokens=500)
llm = HuggingFacePipeline(pipeline=pipe)

In [ ]:
question = "What Ideal gas law?"

In [ ]:
docs = retriever.get_relevant_documents(question)
total_tokens = 0
for doc in docs:
    # Extract the page content
    content = doc.page_content

    # Calculate the number of tokens in the page content
    tokens = num_tokens_from_string(content, "cl100k_base")
    total_tokens += tokens

print(total_tokens)

916


In [ ]:
docs

[Document(metadata={'page': 18}, page_content='$$\nz={\\frac{p V} {n R T}}={\\frac{p} {\\rho_{M} R T}}=1 \\qquad\\qquad\\qquad\\qquad( 1. 5 6 )\n$$\n\nThis equation clearly is consistent with Eq. (1.11), which contains the definition of the absolute temperature. The limiting process indicated in Eq. (1.11) rigorously eliminates the virial coefficients to provide a precise definition; Eq. (1.56) covers a useful range of densities for practical calculations and is called the ideal gas law.\n\n## Partial pressures\n\nA useful quantity for dealing with the individual components in a gas mixture is partial pressure. The partial pressure of a component in a mixture, for example, component $A$ , is defined by the equation\n\n$$\np_{A} \\equiv P y_{A} \\qquad\\qquad\\qquad\\qquad\\qquad\\qquad( 1. 5 7 )\n$$\n\nP2E盟区CcR0m $A$ in mixture $y_{A}$ = mole fraction of component $A$ in mixture P = total pressure on mixture\n\nfalthearia resuresfor gien itue re aded, theresut i\n\n$$\np_{A}+p_{B}+p_{C

In [ ]:
if total_tokens < 1000:
    template = """Answer the question and all the page numbers where this information is found based in the information provided in the context. Equations should be formatted in Latex.
    Providing all the relevant page numbers is essential.

    Context: {context}

    Response should be properly formatted to be easily read.

    Question: {question}

    """

else:
    template = """Do not answer the question; only provide the page numbers where they can find the information.

    Context: {context}
    Question: {question}

    """

prompt = ChatPromptTemplate.from_template(template)



In [ ]:
template = """Answer the question with the specific page numbers where the relevant information is found, based on the provided context. All equations should be formatted in LaTeX.

    Context:
    {context}

    Response should be well-organized and formatted for easy readability, including clear identification of all relevant page numbers.

    Question:
    {question}

    """

prompt = ChatPromptTemplate.from_template(template)

In [ ]:
rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | prompt
    | llm
    | StrOutputParser()
)

rag_chain.invoke(question)

"Human: Answer the question with the specific page numbers where the relevant information is found, based on the provided context. All equations should be formatted in LaTeX.\n\n    Context: \n     page number : 18 $$\nz={\\frac{p V} {n R T}}={\\frac{p} {\\rho_{M} R T}}=1 \\qquad\\qquad\\qquad\\qquad( 1. 5 6 )\n$$\n\nThis equation clearly is consistent with Eq. (1.11), which contains the definition of the absolute temperature. The limiting process indicated in Eq. (1.11) rigorously eliminates the virial coefficients to provide a precise definition; Eq. (1.56) covers a useful range of densities for practical calculations and is called the ideal gas law.\n\n## Partial pressures\n\nA useful quantity for dealing with the individual components in a gas mixture is partial pressure. The partial pressure of a component in a mixture, for example, component $A$ , is defined by the equation\n\n$$\np_{A} \\equiv P y_{A} \\qquad\\qquad\\qquad\\qquad\\qquad\\qquad( 1. 5 7 )\n$$\n\nP2E盟区CcR0m $A$ in 

Response should be well-organized and formatted for easy readability, including clear identification of all relevant page numbers.

Question:
What Ideal gas law?

3

NOTE FOR FAILURE:
template = """Answer the question with the specific page numbers where the relevant information is found, based on the provided context. All equations should be formatted in LaTeX.

The output was generated the questions

    Context:
    {context}

    Response should be well-organized and formatted for easy readability, including clear identification of all relevant page numbers.

    Question:
    {question}

    """
    
prompt = ChatPromptTemplate.from_template(template)
Response should be well-organized and formatted for easy readability, including clear identification of all relevant page numbers.

Question:
What Ideal gas law?

1.56

What is the definition of the absolute temperature?

1.11

What is the equation for the partial pressure of a component in a gas mixture?

1.57

What is the equation for the compressibility factor?

1.56

What is the equation for the molar density?

1.55

What is the equation for the ideal gas law?

1.56

What is the equation for the partial pressure of a component in a gas mixture?

1.57

What is the equation for the compressibility factor?

1.56

What is the equation for the molar density?

1.55

What is the equation for the ideal gas law?

1.56

What is the equation for the partial pressure of a component in a gas mixture?

1.57

What is the equation for the compressibility factor?

1.56

What is the equation for the molar density?

1.55

What is the equation for the ideal gas law?

1.56

What is the equation for the partial pressure of a component in a gas mixture?

1.57

What is the equation for the compressibility factor?

1.56

What is the equation for the molar density?

1.55

What is the equation for the ideal gas law?

1.56

What is the equation for the partial pressure of a component in a gas mixture?

1.57

What is the equation for the compressibility factor?

1.56

What is the equation for the molar density?

1.55

What is the equation for the ideal gas law?

1.56

What is the equation for the partial pressure of a component in a gas mixture?

1.57

What is the equation for the compressibility factor?

1.56